go/database/mpt/archive_trie.go

```diff
@@ -352,6 +412,80 @@ func (a *ArchiveTrie) Close() error {
 		a.head.closeWithError(a.Flush()))
 }
 
+func (a *ArchiveTrie) createCheckpoint() error {
+	// Before the checkpoint can be created, all data needs
+	// to be flushed to the underlying storage.
+	if err := a.Flush(); err != nil {
+		return err
+	}
+	// The creation of the checkpoint makes the current
+	// state recoverable in case of a crash.
+	_, err := a.checkpointCoordinator.CreateCheckpoint()
+	return err
+}
+
+func GetCheckpointBlock(dir string) (uint64, error) {
+	checkpointDir := filepath.Join(dir, fileNameArchiveCheckpointDirectory)
+	coordinator, err := checkpoint.NewCoordinator(checkpointDir)
+	if err != nil {
+		return 0, err
+	}
+	cp := coordinator.GetCurrentCheckpoint()
+	restorer := getRootListRestorer(dir)
+	numRoots, err := restorer.getNumRootsInCheckpoint(cp)
+	return uint64(numRoots - 1), err
+}
+
+func RestoreBlockHeight(directory string, config MptConfig, block uint64) (err error) {
+
+	// Make sure access to the directory is exclusive.
+	lock, err := LockDirectory(directory)
+	if err != nil {
+		return fmt.Errorf("failed to get exclusive access to directory: %v", err)
+	}
+	defer lock.Release()
+
+	// Check available block height -- stop recovery if there are not enough blocks.
+	checkpointHeight, err := GetCheckpointBlock(directory)
+	if err != nil {
+		return fmt.Errorf("failed to get checkpoint height: %v", err)
+	}
+	if block > uint64(checkpointHeight) {
+		return fmt.Errorf("block %d is beyond the last checkpoint height of %d", block, checkpointHeight)
+	}
+
+	// Mark this directory as dirty at least for the duration of the recovery.
+	if err := markDirty(directory); err != nil {
+		return fmt.Errorf("failed to mark directory %s as dirty: %w", directory, err)
+	}
```

Comment:
Shouldnt it already be marked as `dirty`, since at this point, we are restoring it?

Reply:
This can not be known. There is nothing that checks whether the given directory covers indeed a corrupted Archive. It could be a healthy Archive that should just be reverted to an earlier state.